File: src/SharpinoHandlerShim.fs
namespace Sharpino.CommandHandler

open System
open Sharpino.Core
open Sharpino.Storage
open Sharpino.MemoryStorage
open Sharpino.Template.Models

[<AutoOpen>]
module GlobalState =
    let getAggregateStorageFreshStateViewer<'Todo, 'Events, 'Format> (store: MemoryStorage) : AggregateViewer<'Todo> =
        fun (id: Guid) ->
            match states.TryFind id with
            | Some s -> Ok (1, s :?> 'Todo)
            | None -> Error (sprintf "Aggregate not found: %A" id)

    let runInit<'Todo, 'Events, 'Format> (store: IEventStore<'Format>) (senders: obj) (todo: 'Todo) =
        let id =
            match box todo with
            | :? Todo as t -> t.TodoId.Value
            | _ -> Guid.Empty
        states <- states.Add(id, todo)
        Ok ()

    let runAggregateCommand<'Todo, 'Events, 'Format> (id: Guid) (store: IEventStore<'Format>) (senders: obj) (cmd: AggregateCommand<'Todo, 'Events>) =
        match states.TryFind id with
        | Some stateObj ->
            let state = stateObj :?> 'Todo
            match cmd.Execute state with
            | Ok (newState, evts) ->
                states <- states.Add(id, newState)
                let currentEvts = match events.TryFind id with Some e -> e | None -> []
                events <- events.Add(id, currentEvts @ (evts |> List.map box))
                Ok ()
            | Error e -> Error e
        | None -> Error (sprintf "Aggregate not found for command execution: %A" id)

namespace Sharpino.Template

open System
open System.Threading
open System.Threading.Tasks
open Sharpino.Storage
open Sharpino.CommandHandler

module StateView =
    let getAggregateStatesInATimeIntervalAsync<'Todo, 'Events, 'Format> (store: IEventStore<'Format>) (start: DateTime) (finish: DateTime) (ct: Option<CancellationToken>) =
        let allStates =
            states
            |> Map.toList
            |> List.map (fun (k, v) -> (k, v :?> 'Todo))
        #if FABLE_COMPILER
        async { return Ok allStates }
        #else
        Task.FromResult (Ok allStates)
        #endif

// Task and FsToolkit shims for Fable compatibility
[<AutoOpen>]
module TaskShim =
    #if FABLE_COMPILER
    open Fable.Core
    #endif

    type TaskBuilder() =
        #if FABLE_COMPILER
        member this.Bind(x: Task<'T>, f: 'T -> Async<'U>) = async.Bind(Async.AwaitPromise (unbox x), f)
        #else
        member this.Bind(x: Task<'T>, f: 'T -> Async<'U>) = async.Bind(Async.AwaitTask x, f)
        #endif
        member this.Bind(x: Async<'T>, f: 'T -> Async<'U>) = async.Bind(x, f)
        member this.Return(x: 'T) = async.Return(x)
        member this.ReturnFrom(x: Async<'T>) = async.ReturnFrom(x)
        member this.Zero() = async.Zero()
        member this.Delay(f: unit -> Async<'T>) = async.Delay(f)
        member this.Run(f: Async<'T>) = f
    let task = TaskBuilder()

    type TaskResultBuilder() =
        member this.Bind(x: Task<Result<'T, 'E>>, f: 'T -> Async<Result<'U, 'E>>) =
            async {
                #if FABLE_COMPILER
                let! res = Async.AwaitPromise (unbox x)
                #else
                let! res = Async.AwaitTask x
                #endif
                match res with
                | Ok v -> return! f v
                | Error e -> return Error e
            }
        member this.Bind(x: Async<Result<'T, 'E>>, f: 'T -> Async<Result<'U, 'E>>) =
            async {
                let! res = x
                match res with
                | Ok v -> return! f v
                | Error e -> return Error e
            }
        member this.Return(x: 'T) = async.Return(Ok x)
        member this.ReturnFrom(x: Async<Result<'T, 'E>>) = x
    let taskResult = TaskResultBuilder()

// FSharpPlus shims for Fable compatibility
[<AutoOpen>]
module FSharpPlusShim =
    let inline (|>>) x f = List.map f x

// Mock Result.get for Tests.fs
namespace FsToolkit.ErrorHandling
module Result =
    let get = function Ok x -> x | Error e -> failwithf "Expected Ok, got Error: %A" e

// Mock DotNetEnv for Tests.fs
namespace DotNetEnv
module Env =
    let Load() = ()

File: src/SharpinoHandlerShim.js
import { toList, FSharpMap__Add, FSharpMap__TryFind } from "./fable_modules/fable-library.4.1.4/Map.js";
import { Storage_GlobalStorageState_events, Storage_GlobalStorageState_states } from "./SharpinoCoreShim.js";
import { toFail, printf, toText } from "./fable_modules/fable-library.4.1.4/String.js";
import { FSharpResult$2 } from "./fable_modules/fable-library.4.1.4/Choice.js";
import { value as value_1 } from "./fable_modules/fable-library.4.1.4/Option.js";
import { Todo } from "./Models/Todo/Todo.js";
import { TodoId__get_Value } from "./Commons.js";
import { map, append, empty } from "./fable_modules/fable-library.4.1.4/List.js";
import { singleton } from "./fable_modules/fable-library.4.1.4/AsyncBuilder.js";
import { class_type } from "./fable_modules/fable-library.4.1.4/Reflection.js";
import { awaitPromise } from "./fable_modules/fable-library.4.1.4/Async.js";

export function Sharpino_CommandHandler_GlobalState_getAggregateStorageFreshStateViewer(store, id) {
    const matchValue = FSharpMap__TryFind(Storage_GlobalStorageState_states(), id);
    if (matchValue == null) {
        return new FSharpResult$2(1, [toText(printf("Aggregate not found: %A"))(id)]);
    }
    else {
        return new FSharpResult$2(0, [[1, value_1(matchValue)]]);
    }
}

export function Sharpino_CommandHandler_GlobalState_runInit(store, senders, todo) {
    let t;
    let id;
    const matchValue = todo;
    id = ((matchValue instanceof Todo) ? ((t = matchValue, TodoId__get_Value(t.TodoId))) : "00000000-0000-0000-0000-000000000000");
    Storage_GlobalStorageState_states(FSharpMap__Add(Storage_GlobalStorageState_states(), id, todo));
    return new FSharpResult$2(0, [void 0]);
}

export function Sharpino_CommandHandler_GlobalState_runAggregateCommand(id, store, senders, cmd) {
    const matchValue = FSharpMap__TryFind(Storage_GlobalStorageState_states(), id);
    if (matchValue == null) {
        return new FSharpResult$2(1, [toText(printf("Aggregate not found for command execution: %A"))(id)]);
    }
    else {
        const state = value_1(matchValue);
        const matchValue_1 = cmd.Execute(state);
        if (matchValue_1.tag === 1) {
            return new FSharpResult$2(1, [matchValue_1.fields[0]]);
        }
        else {
            Storage_GlobalStorageState_states(FSharpMap__Add(Storage_GlobalStorageState_states(), id, matchValue_1.fields[0][0]));
            let currentEvts;
            const matchValue_2 = FSharpMap__TryFind(Storage_GlobalStorageState_events(), id);
            currentEvts = ((matchValue_2 == null) ? empty() : matchValue_2);
            Storage_GlobalStorageState_events(FSharpMap__Add(Storage_GlobalStorageState_events(), id, append(currentEvts, map((value) => value, matchValue_1.fields[0][1]))));
            return new FSharpResult$2(0, [void 0]);
        }
    }
}

export function Sharpino_Template_StateView_getAggregateStatesInATimeIntervalAsync(store, start, finish, ct) {
    const allStates = map((tupledArg) => [tupledArg[0], tupledArg[1]], toList(Storage_GlobalStorageState_states()));
    return singleton.Delay(() => singleton.Return(new FSharpResult$2(0, [allStates])));
}

export class Sharpino_Template_TaskShim_TaskBuilder {
    constructor() {
    }
}

export function Sharpino_Template_TaskShim_TaskBuilder_$reflection() {
    return class_type("Sharpino.Template.TaskShim.TaskBuilder", void 0, Sharpino_Template_TaskShim_TaskBuilder);
}

export function Sharpino_Template_TaskShim_TaskBuilder_$ctor() {
    return new Sharpino_Template_TaskShim_TaskBuilder();
}

export function Sharpino_Template_TaskShim_TaskBuilder__Bind_Z521FF722(this$, x, f) {
    return singleton.Bind(awaitPromise(x), f);
}

export function Sharpino_Template_TaskShim_TaskBuilder__Bind_7A510B33(this$, x, f) {
    return singleton.Bind(x, f);
}

export function Sharpino_Template_TaskShim_TaskBuilder__Return_1505(this$, x) {
    return singleton.Return(x);
}

export function Sharpino_Template_TaskShim_TaskBuilder__ReturnFrom_ZD4A93B1(this$, x) {
    return singleton.ReturnFrom(x);
}

export function Sharpino_Template_TaskShim_TaskBuilder__Zero(this$) {
    return singleton.Zero();
}

export function Sharpino_Template_TaskShim_TaskBuilder__Delay_Z5276B41B(this$, f) {
    return singleton.Delay(f);
}

export function Sharpino_Template_TaskShim_TaskBuilder__Run_ZD4A93B1(this$, f) {
    return f;
}

export const Sharpino_Template_TaskShim_task = Sharpino_Template_TaskShim_TaskBuilder_$ctor();

export class Sharpino_Template_TaskShim_TaskResultBuilder {
    constructor() {
    }
}

export function Sharpino_Template_TaskShim_TaskResultBuilder_$reflection() {
    return class_type("Sharpino.Template.TaskShim.TaskResultBuilder", void 0, Sharpino_Template_TaskShim_TaskResultBuilder);
}

export function Sharpino_Template_TaskShim_TaskResultBuilder_$ctor() {
    return new Sharpino_Template_TaskShim_TaskResultBuilder();
}

export function Sharpino_Template_TaskShim_TaskResultBuilder__Bind_Z6722B822(this$, x, f) {
    return singleton.Delay(() => singleton.Bind(awaitPromise(x), (_arg) => {
        const res = _arg;
        return (res.tag === 1) ? singleton.Return(new FSharpResult$2(1, [res.fields[0]])) : singleton.ReturnFrom(f(res.fields[0]));
    }));
}

export function Sharpino_Template_TaskShim_TaskResultBuilder__Bind_454FBA93(this$, x, f) {
    return singleton.Delay(() => singleton.Bind(x, (_arg) => {
        const res = _arg;
        return (res.tag === 1) ? singleton.Return(new FSharpResult$2(1, [res.fields[0]])) : singleton.ReturnFrom(f(res.fields[0]));
    }));
}

export function Sharpino_Template_TaskShim_TaskResultBuilder__Return_1505(this$, x) {
    return singleton.Return(new FSharpResult$2(0, [x]));
}

export function Sharpino_Template_TaskShim_TaskResultBuilder__ReturnFrom_2D6B8459(this$, x) {
    return x;
}

export const Sharpino_Template_TaskShim_taskResult = Sharpino_Template_TaskShim_TaskResultBuilder_$ctor();

export function FsToolkit_ErrorHandling_Result_get(_arg) {
    if (_arg.tag === 1) {
        return toFail(printf("Expected Ok, got Error: %A"))(_arg.fields[0]);
    }
    else {
        return _arg.fields[0];
    }
}

export function DotNetEnv_Env_Load() {
}

